ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/Interceptions.kt

```diff
@@ -0,0 +1,165 @@
+ * It is a wrapper over pipeline.intercept(phase) { ... } and is needed to hide old Plugins API functionality
+ * */
+public class Interception<T : Any>(
+    public val phase: PipelinePhase,
+    public val action: (Pipeline<T, ApplicationCall>) -> Unit
+)
+
+/**
+ * Compatibility class. Interception class for Call phase
+ * */
+public typealias CallInterception = Interception<Unit>
+
+/**
+ * Compatibility class. Interception class for Receive phase
+ * */
+
+public typealias ReceiveInterception = Interception<ApplicationReceiveRequest>
+
+/**
+ * Compatibility class. Interception class for Send phase
+ * */
+public typealias ResponseInterception = Interception<Any>
+
+/**
+ * Compatibility class. Every plugin that needs to be compatible with `beforePlugin(...)` and `afterPlugin(...)`
+ * needs to implement this class. It defines a list of interceptions (see [Interception]) for phases in different pipelines
+ * that are being intercepted by the current plugin.
+ *
+ * It is needed in order to get first/last phase in each pipeline for the current plugin and create a new phase
+ * that is strictly earlier/later in this pipeline than any interception of the current plugin.
+ *
+ * Note: any [KtorPlugin] instance automatically fills these fields and there is no need to implement them by hand.
+ * But if you want to use old plugin API based on pipelines and phases and to make it available for `beforePlugin`/
+ * `afterPlugin` methods of the new plugins API, please consider filling the phases your plugin defines by implementing
+ * this interface [InterceptionsHolder].
+ *
+ * Please, use [defineInterceptions] method inside the first init block to simply define all the interceptions for your
+ * feature. Also it is useful to always delegate to [DefaultInterceptionsHolder] for simplicity.
+ * */
+public interface InterceptionsHolder {
+    /**
+     * A name for the current plugin.
+     * */
+    public val name: String get() = this.javaClass.simpleName
+
+    public val key: AttributeKey<out InterceptionsHolder>
+
+    @InternalAPI
+    public val fallbackInterceptions: MutableList<CallInterception>
+
+    @InternalAPI
+    public val callInterceptions: MutableList<CallInterception>
+
+    @InternalAPI
+    public val monitoringInterceptions: MutableList<CallInterception>
+
+    @InternalAPI
+    public val beforeReceiveInterceptions: MutableList<ReceiveInterception>
+
+    @InternalAPI
+    public val onReceiveInterceptions: MutableList<ReceiveInterception>
+
+    @InternalAPI
+    public val beforeResponseInterceptions: MutableList<ResponseInterception>
+
+    @InternalAPI
+    public val onResponseInterceptions: MutableList<ResponseInterception>
+
+    @InternalAPI
+    public val afterResponseInterceptions: MutableList<ResponseInterception>
+
+    public fun newPhase(): PipelinePhase = PipelinePhase("${name}Phase${Random.nextInt()}")
```

Comment:
why do we need a method for the constructor?

Please avoid using random in such cases